esm: merge esm and cjs package.json caches

shackijj opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->
Relates to: #30674
##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

shackijj commented:
@guybedford as discussed earlier in the issue, I tried to keep diffs down. 
As the next step, we could move `readPackageScope`, `readPackage`, `packageJsonCache` functions from cjs loader to a separate module. The module will be used by esm and cjs loaders.

Additionally, I'd move the following checks to `readPackage` function (previously they were in getPackageConfig in esm loader).
```
  if (typeof main !== 'string') main = undefined;	
  if (typeof name !== 'string') name = undefined;	
  // Ignore unknown types for forwards compatibility	
  if (type !== 'module' && type !== 'commonjs') type = 'none'
```

shackijj commented:
One test failed on MacOS.
```
not ok 2356 parallel/test-worker-init-failure
  ---
  duration_ms: 2.77
  severity: fail
  exitcode: 1
  stack: |-
    child stdout: 
    
    child stderr: assert.js:920
        throw err;
        ^
    
    AssertionError [ERR_ASSERTION]: The input did not match the regular expression /EMFILE/. Input:
    
    'ENOENT: no such file or directory, uv_cwd'
    
        at Worker.<anonymous> (/Users/runner/runners/2.169.1/work/node/node/node-v15.0.0-nightly2020-05-04xxxx/test/parallel/test-worker-init-failure.js:38:14)
        at Worker.emit (events.js:315:20)
        at Worker.[kOnErrorMessage] (internal/worker.js:233:10)
        at Worker.[kOnMessage] (internal/worker.js:243:37)
        at MessagePort.<anonymous> (internal/worker.js:164:57)
        at MessagePort.emit (events.js:315:20)
        at MessagePort.onmessage (internal/worker/io.js:78:8)
        at MessagePort.exports.emitMessage (internal/per_context/messageport.js:11:10)
```
It seems that the test does not relate the changes I made. Also, this [job](https://github.com/nodejs/node/pull/33229/checks?check_run_id=642832747) finished succefully (same MacOS environment). Can't reproduce this on my Mac as well.
`test/parallel/test-worker-init-failure.js` checks extreme scenario by creating 256 workers.

Is there a chance that the test is flaky? Is there a way to restrart ci pipeline without a new commit?

@guybedford do you have any thought on that?

guybedford commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -552,28 +450,28 @@ function packageResolve(specifier, base, conditions) {
-  const packageConfig = getPackageScopeConfig(base, base);
-  if (packageConfig.exists) {
+  const packageScope = readPackageScope(fileURLToPath(base));
+  if (packageScope !== false) {
```
Perhaps just define `const packageConfig = packageScope.data` here.

guybedford commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -595,13 +493,15 @@ function packageResolve(specifier, base, conditions) {
     // Package match.
-    const packageConfig = getPackageConfig(packageJSONPath, base);
+    const packagePath = StringPrototypeReplace(
+      packageJSONPath, `${sep}package.json`, '');
```
I believe this will actually replace eg `/path/to/package.json/package.json` with `/path/to//package.json`.

Rather just trim the end off since it's guaranteed to end in `/package.json` - `packagePath.slice(0, -'/package.json'.length)` sort of thing.

guybedford commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -76,107 +75,6 @@ function tryStatSync(path) {
-  const existing = packageJSONCache.get(path);
-  if (existing !== undefined) {
-    if (!existing.isValid) {
-      throw new ERR_INVALID_PACKAGE_CONFIG(path, fileURLToPath(base), false);
```
I'd have expected removing this would have caused a failure, but it seems we don't have a test for invalid JSON syntax in the package.json.

Can we please add one to this PR? Along the lines of -

test/fixtures/node_modules/invalid-pjson/package.json:
```js
{
  "invalid"
}
```

Where the test case then tries to load the package, which should throw an `ERR_INVALID_PACKAGE_CONFIG` error.

The CommonJS failure may be different here and I suppose that divergence is ok since CommonJS should throw as it historically has, but the ESM error is supposed to be the `ERR_INVALID_PACKAGE_CONFIG` by the current spec and this code path.

guybedford commented:
@shackijj those further refactorings sound really sensible to me too, if you want to add them in this PR too feel free.

shackijj commented:
@guybedford I decided not to move the functions to a separate module until you approve the last changes that I made.

Added test for ERR_INVALID_PACKAGE_CONFIG.

`internalModuleReadJSON` maps an invalid JSON to a valid JSON (e.g `}{` -> `{}`, `{'invalid'}` -> `{}`), so we need a different file reading strategy in ESM loader as to throw ERR_INVALID_PACKAGE_CONFIG.

P.S. I believe we could document internalModuleReadJSON a bit better.

guybedford commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -584,7 +527,7 @@ function packageResolve(specifier, base, conditions) {
   do {
     const stat = tryStatSync(
-      StringPrototypeSlice(packageJSONPath, 0, packageJSONPath.length - 13));
+      dirname(packageJSONPath));
```
I would actually prefer we stick with slicing just because it's quite well-defined what string manipulation steps are involved.

guybedford commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -506,9 +408,49 @@ function packageExportsResolve(
+}
+
+function catchInvalidPackage(error, path, base) {
+  if (StringPrototypeIncludes(error.message, 'Error parsing')) {
```
We could do an `error instanceof SyntaxError` here since the JSON error is the only case of a syntax error that can happen on the code path.

guybedford commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -506,9 +408,49 @@ function packageExportsResolve(
+function catchInvalidPackage(error, path, base) {
+  if (StringPrototypeIncludes(error.message, 'Error parsing')) {
+    throw new ERR_INVALID_PACKAGE_CONFIG(
+      join(path, 'package.json'), fileURLToPath(base), false);
```
We should carry through the parse message from `error.message` as well here for context (`error.message.slice(('Error parsing ' + path + ': ').length)` sort of thing maybe)

guybedford commented on test/es-module/test-esm-invalid-pjson.js:
```diff
@@ -0,0 +1,29 @@
+      [
+        '[ERR_INVALID_PACKAGE_CONFIG]: ',
+        `Invalid package config ${invalidJson}, `,
+        'Unexpected token } in JSON at position 14'
```
It seems this becomes `position 16` in windows resulting in the Windows test failure.

```suggestion
        `Unexpected token } in JSON at position ${isWin ? 16 : 14}`
```

shackijj commented:
@guybedford I rebased the branch and changed the history a bit as to follow the commit guideline. Also, the guideline states that I should add 'author ready' label to this PR. It seems that I don't have rights to add a label to PR. Could you please add it?

jkrems commented on test/fixtures/node_modules/invalid-pjson/package.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "invalid"
+}
```
Should we still have the trailing new line here?

shackijj replied:
Is there a rule somewhere in guidelines about trailing new lines?
Linter doesn't fail, so we could leave it as-is. I don't mind to add the trailing new line though.

jkrems replied:
It's a general thing that will show up in the Github UI as a red marker and will often be auto-fixed by editors. Having no trailing new line is usually only done if there's a specific reason for it. Given that this file is called "invalid-pjson/package.json", it may direct people's attention to the wrong thing (the missing newline) instead of the thing that's actually invalid (no property value).

jkrems commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -239,16 +241,22 @@ Module._debug = deprecate(debug, 'Module._debug is deprecated.', 'DEP0077');
   if (existing !== undefined) return existing;
 
-  const json = internalModuleReadJSON(path.toNamespacedPath(jsonPath));
+  const json = readJsonStrategy(jsonPath);
```
This looks like we're using different functions to populate the same cache key. Is there any danger of race conditions where one strategy writes a key and another then silently gets the value?

shackijj replied:
Goog point! 

> Is there any danger of race conditions where one strategy writes a key and another then silently gets the value?

The worst scenario is the following. CJS reads invalid package.json and puts in the cache `{}`. ESM module then reads the `{}` and tries to work with it, although ESM should throw error in this case.

I believe it would be better to use `internalModuleReadJSON` for both ESM and CJS. First, It allows to use the same cache. Second, `internalModuleReadJSON` is resilient to failure.
We would need to sacrifice nice error reporting for ESM. But we could explicitly describe this behavior (the case when package.json is invalid) in docs and it won't be a problem.

 @guybedford what do you think?

jkrems replied:
Could we make it always return the same `{}` and expose it as `INVALID_PACKAGE_JSON`? That way ESM could check for strict equality and maybe still throw..?

guybedford replied:
Yes this does sound like a problem. Thanks @jkrems for spotting this.

This does sound like a serious inconsistency with sharing the caches, with the only solutions that aren't unsound being:

1. Change the ESM resolver to pass on invalid JSON without these fields, like this does for the CJS resolver.
2. Fixup `internalModuleReadJSON` so that it's parser actually detects invalid JSON state.

I suspect (2) is probably the best approach moving forward here to be honest.

@trevnorris we may need to create a new issue here to track the fact that `internalModuleReadJSON` will not throw JSON parsing errors for invalid syntax JSON files without the search keywords in them. Do you think a fix to ensure invalid JSON parser state is comprehensively detected to return `undefined` would be straightforward.

@shackijj if you're proficient with C++ it might be worth looking into as well.

guybedford replied:
@bnoordhuis we may need to create a new issue here to track the fact that `internalModuleReadJSON` will not throw JSON parsing errors for invalid syntax JSON files without the search keywords in them. Do you think a fix to ensure invalid JSON parser state is comprehensively detected to return `undefined` would be straightforward?

Apologies @trevnorris for copying you in by mistake :)

shackijj replied:
@guybedford CJS loader detects invalid JSON state [here](https://github.com/nodejs/node/blob/8a6fab02adab2de05f6e864847f96b0924be0840/lib/internal/modules/cjs/loader.js#L263). I think that parsing JSON in `internalModuleReadJSON` might lead to performance issues unless there is a way to parse string there (in `internalModuleReadJSON`) and get a reference to a resulting (parsed) object in JS (i'm not sure if it's possible).

Alternatively, we might create a cache that will store only package.json strings (not parsed objects as it's now). ESM and CJS could parse the string the way they do now and have their own caches that will store parsed objects.

![image](https://user-images.githubusercontent.com/9201657/81545819-22607600-9382-11ea-97d8-8c3ab3f7f601.png)

guybedford replied:
@shackijj the specific issue is a package.json file like `{ "invalid" }` is detected by `internalModuleReadJSON` as not containing properties `"main"`, `"exports"`, `"type"` or `"name"` so that function returns `"{}"` - the empty object string, instead of invalid. So CommonJS isn't currently throwing on these files.

`internalModuleReadJSON` is already a rudimentary JSON parser. It doesn't seem like it would be a big stretch to give it the ability itself in the parsing loop to detect invalid syntax and explicitly return `undefined` in these cases to indicate the error.

shackijj replied:
>  It doesn't seem like it would be a big stretch to give it the ability itself in the parsing loop to detect invalid syntax and explicitly return undefined in these cases to indicate the error.

It's possible, but it will significantly change the behaviour of CJS (seems like breaking a change to me). Second, I think, that we will need to reimplement JSONParse in `internalModuleReadJSON`.

jkrems replied:
It feels like an extreme edge case to me that there's a `package.json` that both has syntax errors _and_ doesn't include any of the relevant metadata fields. Should we just accept the CJS behavior for ESM here? The fallout seems extremely limited.

guybedford replied:
Yes it is an edge case, but the reason we have a resolver specification that is public is because we aim to define all the edge cases comprehensively so that other projects can implement resolver logic that matches Node.js exactly.

Adding a line to the resolver specification like "if the JSON does not contain the strings main, name, type or exports then do not throw any invalid JSON syntax errors" seems like a very bad idea to me (within the edge case of course) as it's highly inconsistent. 

@shackjj I agree the CJS case would be breaking to change, but in theory it was also breaking when it was first added. There is actually another way we could do this actually. Instead of returning `emptyObjectString` in the `internalModuleReadJSON` function, we could instead return an object like `{ source: string | undefined, containsKeys: boolean }`. This would then split the function down into three cases:

1. `source: string, containsKeys: true`: String for JSON source when it has the keys (same between esm and cjs loaders)
2. `source: undefined, containsKeys: any`: Missing file, same between esm and cjs loaders
3. `source: string, containsKeys: false`: For CJS source it just treats it as am empty object. For ESM it would still parse the source to return any parse error if necessary.

shackijj replied:
> Adding a line to the resolver specification like "if the JSON does not contain the strings main, name, type or exports then do not throw any invalid JSON syntax errors" seems like a very bad idea to me (within the edge case of course) as it's highly inconsistent.

Agree

> Instead of returning string | undefined from the internalModuleReadJSON function, we could instead return an object like { source: string | undefined, containsKeys: boolean }

Good idea! I need to read about C++/JS communication in node because I haven't worked with v8 C++ API before. Will try to submit a patch during this week.

@bnoordhuis Could you please share you thoughts on @guybedford's suggestion?

guybedford replied:
Sounds good - if it helps, you should be able to create an object something like https://github.com/nodejs/node/blob/master/src/node_process_object.cc#L93 and set properties something like https://github.com/nodejs/node/blob/master/src/node_process_object.cc#L157. I'm terrible at this stuff so don't quote me :)

ljharb commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -117,17 +94,9 @@ function getPackageConfig(path, base) {
-    packageJSONCache.set(path, packageConfig);
-    return packageConfig;
+  } catch (error) {
+    const errorPath = StringPrototypeSlice(path, 0, path.length - 13);
```
what is this magic number 13 for?

targos replied:
I guess `'package.json'.length - 1`

ljharb replied:
a comment to that effect may help then :-)

guybedford commented:
This looks great to me. //cc @addaleax for C++ review.

addaleax commented on src/node_file.cc:
```diff
@@ -912,18 +916,18 @@ static void InternalModuleReadJSON(const FunctionCallbackInfo<Value>& args) {
+  return_value->Set(
+    isolate->GetCurrentContext(),
+    FIXED_ONE_BYTE_STRING(isolate, "containsKeys"),
+    Boolean::New(isolate, p < pe ? true : false)).Check();
```
If the code here is kept-as is:

The `FIXED_ONE_BYTE_STRING()`s should be listed in the table in `env.h` instead and used as e.g. `env->contains_keys_string()`, because this code runs more than a handful of times during a Node.js process and that way we can avoid creating the strings over and over again.

Also, instead of `.Check()`s here we should be doing proper error handling, i.e. `if (return_value->Set(...).IsNothing()) return;`.

---

All that being said: The `return_values->Set()` calls will have a non-trivial performance impact (I don’t know how relevant that is here, but given that this function was introduced as an optimization, I assume that it is). A nicer alternative might be to return an `Array` from this C++ binding function with 2 entries, e.g.:

```c++
  Local<Value> return_value[] = {
    String::NewFromUtf8(...).ToLocalChecked(),
    Boolean::New(...)
  }
  args.GetReturnValue().Set(
      Array::New(isolate, return_value, arraysize(return_value)));
```

The downside is that this would not read quite as nicely from JS, but that could be addressed by destructuring in the wrapper module, I think:

```js
  const [ string, containsKeys ] = internalModuleReadJSON(path);
  cache.set(path, { string, containsKeys });
```

I think the rest of the code should keep working as it does in that case.

shackijj commented:
The tests failed on MacOS with the following error
```
AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
+ actual - expected

+ 'ERR_STREAM_DESTROYED'
- 'ECONNRESET'
    at ClientRequest.<anonymous> (/Users/runner/runners/2.262.1/work/node/node/test/parallel/test-http-destroyed-socket-write2.js:66:16)
    at ClientRequest.<anonymous> (/Users/runner/runners/2.262.1/work/node/node/test/common/index.js:363:15)
    at ClientRequest.emit (events.js:315:20)
    at emitErrorNt (_http_outgoing.js:666:43)
    at processTicksAndRejections (internal/process/task_queues.js:85:21) {
  generatedMessage: true,
  code: 'ERR_ASSERTION',
  actual: 'ERR_STREAM_DESTROYED',
  expected: 'ECONNRESET',
  operator: 'strictEqual'
}
Command: out/Release/node /Users/runner/runners/2.262.1/work/node/node/test/parallel/test-http-destroyed-socket-write2.js
```

It seems that the test is not related this fix and might be flaky.

GeoffreyBooth commented:
https://ci.nodejs.org/job/node-test-pull-request/31456/ is green, this can land.

GeoffreyBooth commented:
Landed in 8f10bb2da5bc.
